feat: Rebase `feature/optional` to `develop`

fsandhei opened the pull request:
Rebasing `feature/optional` in attempt to revive this PR, referring to https://github.com/nlohmann/json/pull/2117.

Sorry for long time between responses.

I may have been going forward with this the "wrong" way, so I apologize in advance.

Please let me know if this should be handled differently.

## TODO
There is one test that is failing: Conversions from a default initialized `nlohmann::json` to `std::optional<T>` seems to fail. In the test case below, the last assertion fails with an exception. This is tested on Arch Linux with GCC 13.1.1.

I was not able to figure out how to fix this so I need some assistance here.

```
/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573:
TEST CASE:  std::optional
  null

/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573: ERROR: test case THREW exception: exception thrown in subcase - will translate later when the whole test case has been exited (cannot translate while there is an active exception)

===============================================================================
/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573:
TEST CASE:  std::optional

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  null

/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573: ERROR: test case THREW exception: [json.exception.type_error.302] type must be string, but is null
```

nlohmann committed:
:construction: conversions for std::optional

nlohmann committed:
:checkered_flag: fix <optional> inclusion

nlohmann committed:
:green_heart: overwork tests

remedi committed:
Use JSON_HAS_CPP_17 only after it has been defined

nlohmann committed:
:white_check_mark: update tests

nlohmann committed:
:checkered_flag: include right <optional> header

nlohmann committed:
:recycle: do not include experimental headers

fsandhei committed:
Add missing #endif after rebase

fredriksd commented:
Think something got messed up on comparison, so closing to not make noise. Sorry about that.

fsandhei commented:
Changed the base branch to `develop`.

I'm sorry if I'm making it a bit difficult here; I forked the repository and rebased `feature/optional` to `develop` and created this PR against `nlohmann:feature/develop` but I noticed that included 590 commits (which makes sense) but it cluttered the actual change here, so I instead changed it towards `nlohmann:develop`.

lightyear15 commented:
I am also interested in having ``std::optional`` support landing in future versions of this library.
I see there was a lot of discussion on how to properly support and convert c++ optional fields vs json nullable fields.
Is there a final decision on the matter? How can we revive this PR and push forward support of ``std::optional`` fields?
@nlohmann

lightyear15 commented on include/nlohmann/detail/conversions/from_json.hpp:
```diff
@@ -43,6 +43,21 @@ inline void from_json(const BasicJsonType& j, typename std::nullptr_t& n)
+    }
+    else
+    {
+        opt = j.template get<T>();
```
```suggestion
        opt = std::make_optional(j.template get<T>());
```

lightyear15 commented on tests/src/unit-conversions.cpp:
```diff
@@ -1569,4 +1569,70 @@ TEST_CASE("JSON to enum mapping")
+        std::optional<std::string> opt_string = "string";
+
+        CHECK(json(opt_string) == j_string);
+        CHECK(std::optional<std::string>(j_string) == opt_string);
```
```suggestion
        CHECK(j_null.get<std::optional<std::string>>() == opt_string);
```

fsandhei committed:
Fix failing test

fsandhei committed:
Only define conversion to std::optional when JSON_USE_IMPLICIT_CONVERSION is disabled.

fsandhei commented:
See that clang-tidy is failing on some checks. Apparently this is on develop, too. I see that it is related to a supposed new check from clang-tidy v19: https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html.

nlohmann commented:
> See that clang-tidy is failing on some checks. Apparently this is on develop, too. I see that it is related to a supposed new check from clang-tidy v19: [clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html).

I'm on it. May take some more time.

lightyear15 commented:
has this PR stalled again?

fsandhei commented:
I'm currently blocked by https://github.com/nlohmann/json/pull/4311.

fsandhei commented:
There has not been any activity in #4311 in a while. I wonder if this would be okay to merge?

fsandhei committed:
missing endif

fsandhei committed:
Remove Wfloat-equal suppress

fsandhei committed:
amalgamate

fsandhei committed:
Move include of optional out of macro_scope; probably does not make sense to be there

nlohmann commented:
Thanks for keeping this alive. I will have a look at this on Saturday.

fsandhei committed:
Make clang-tidy happy

fsandhei commented:
> Thanks for keeping this alive. I will have a look at this on Saturday.

Happy to help! Hopefully this can go in soon.

nlohmann commented:
Please suppress the last clang-tidy warning and not change the code.

fsandhei committed:
Suppress lint instead of changing to 'contains'

fsandhei commented:
> Please suppress the last clang-tidy warning and not change the code.

Reverted change with [296549c](https://github.com/nlohmann/json/pull/4036/commits/296549c3ad3941ccf007db543c8e0d3f908a67d0).

nlohmann reviewed (approved):
Looks good to me.

nlohmann commented:
Thanks!
